ADO.NET/OAuthSampleApp/OAuthSampleApp.cs

```diff
@@ -71,13 +71,35 @@ public static async Task TearDown()
     // uses the password grant flow
     static async Task EnsureAccessToken()
     {
+        string refreshToken = Environment.GetEnvironmentVariable(OAUTH_REFRESH_TOKEN_VAR_STRING, EnvironmentVariableTarget.User);
         string accessToken = Environment.GetEnvironmentVariable(OAUTH_ACCESS_TOKEN_VAR_STRING, EnvironmentVariableTarget.User);
 
         if (string.IsNullOrEmpty(accessToken))
         {
-            // Obtain first access token
-            Console.WriteLine("Access token not found. Obtaining first access token from IDP.");
-            await GetTokensByPasswordGrant();
+            if (string.IsNullOrEmpty(refreshToken))
+            {
+                // Obtain first access token
+                Console.WriteLine("Access token not found. Obtaining first access token from IDP.");
+                await GetTokensByPasswordGrant();
+            }
+            else
+            {
+                try
+                {
+                    // attempt to use given refresh token
+                    Console.WriteLine("Attempting to use given refresh token.");
+                    await GetTokensByRefreshGrant();
+                }
+                catch (HttpRequestException hre)
+                {
+                    Console.WriteLine(hre.Message);
+                    Console.WriteLine("Initial refresh token has expired. Getting new access and refresh tokens.");

```

Comment:
Is this too specific? I believe this would get thrown by the EnsureSuccessStatusCode in getAndSetTokens. What if it's just a bad url, like the user provided a malformed token url? In that case we could just get a 400 bad request response, it would throw an HttpRequestException and then here we would assume that it is an expired refresh token, log that information to the user, and try again with the same bad url. 

Should we specifically check for a status code like 401 unauthorized for expired refresh token?

Reply:
I fixed up the error messages so there isn't a whole cluster of them when there's an exception. There's specific ones for bad client secret/id and bad token urls now. I